src/Core/src/HttpClient/RetryResponse.php

```diff
@@ -0,0 +1,208 @@
+    {
+        $wrappedResponses = [];
+        $map = new \SplObjectStorage();
+        $client = null;
+
+        foreach ($responses as $r) {
+            if (!$r instanceof self) {
+                throw new \TypeError(sprintf('"%s::stream()" expects parameter 1 to be an iterable of AsyncResponse objects, "%s" given.', static::class, get_debug_type($r)));
+            }
+
+            if (null === $client) {
+                $client = $r->client;
+            } elseif ($r->client !== $client) {
+                throw new TransportException('Cannot stream AsyncResponse objects with many clients.');
+            }
+        }
+
+        if (!$client) {
+            return;
+        }
+
+        $end = null;
+        if (null !== $timeout) {
+            $end = \microtime(true) + $timeout;
+        }
+
+        while (true) {
+            foreach ($responses as $r) {
+                $wrappedResponses[] = $r->inner;
+                $map[$r->inner] = $r;
+            }
+            $subTimeout = null;
+            if (null !== $end && ($subTimeout = $end - \microtime(true)) <= 0) {
+                foreach ($map as $response) {
+                    yield $response => new ErrorChunk(0, sprintf('Idle timeout reached for "%s".', $response->getInfo('url')));
+                }
+
+                break;
+            }
+
+            foreach ($client->stream($wrappedResponses, $subTimeout) as $response => $chunk) {
+                /** @var self $r */
+                $r = $map[$response];
+                if (!$chunk->isTimeout() && $chunk->isFirst()) {
+                    if ($r->handleRetry()) {
+                        continue 2;
+                    }
+                }
+                yield $r => $chunk;
+            }
+
+            break;
+        }
+    }
+
+    private function initialize(): void
+    {
+        while (!$this->initialized) {
+            $this->handleRetry();
+        }
+    }
+
+    /**
+     * @return bool return true when the request have been retried
+     *
+     * @throws TransportExceptionInterface
+     */
+    private function handleRetry(): bool
+    {
+        if ($this->initialized) {
+            return false;
+        }
+
+        try {
+            if (($status = $this->inner->getStatusCode()) >= 500) {
+                if (++$this->tryCount <= $this->maxTryCount) {
+                    $this->logger->info('HTTP request failed with status code {statusCode}. Retry the request {attempt}/{maxAttempts}.', [
+                        'statusCode' => $status,
+                        'attempt' => $this->tryCount,
+                        'maxAttempts' => $this->maxTryCount,
+                    ]);
+                    $this->inner = $this->client->request($this->method, $this->url, $this->options);
+
+                    return true;
+                }
+
+                $this->logger->info('HTTP request failed with status code {statusCode}. Stop after {maxAttempts} attempts.', [
```

Comment:
Is anything else logged or is this the only message you see?

We should maybe consider to make this an “error”.

Reply:
good idea